syntax = "proto3";

package ozonmp.bss_office_facade.v1;

import "validate/validate.proto";
import "google/api/annotations.proto";
import "google/protobuf/timestamp.proto";

option go_package = "github.com/ozonmp/bss-office-facade/pkg/bss-office-facade;bss_office_facade";

// BssOfficeApiService - Service for working with offices
service BssOfficeApiService {
  // DescribeOfficeV1 - Describe a office
  rpc DescribeOfficeV1(DescribeOfficeV1Request) returns (DescribeOfficeV1Response) {
    option (google.api.http) = {
      get: "/api/v1/office/{office_id}"
    };
  }
  // CreateOfficeV1 - Create new office
  rpc CreateOfficeV1(CreateOfficeV1Request) returns (CreateOfficeV1Response) {
    option (google.api.http) = {
      post: "/api/v1/office",
      body: "*"
    };
  }

  // RemoveOfficeV1 - delete the office by id
  rpc RemoveOfficeV1(RemoveOfficeV1Request) returns (RemoveOfficeV1Response) {
    option (google.api.http) = {
      delete: "/api/v1/office/{office_id}"
    };
  }

  // ListOfficeV1 - list of offices
  rpc ListOfficesV1(ListOfficesV1Request) returns (ListOfficesV1Response) {
    option (google.api.http) = {
      get: "/api/v1/office/list/{limit}/{offset}"
    };
  }

  // UpdateOfficeV1 - update office fields
  rpc UpdateOfficeV1(UpdateOfficeV1Request) returns (UpdateOfficeV1Response) {
    option (google.api.http) = {
      put: "/api/v1/office",
      body: "*"
    };
  }
}

message Office {
  uint64 id = 1 [(validate.rules).uint64.gt = 0];
  string name = 2  [(validate.rules).string = {min_len: 2, max_len: 100}];
  string description = 3;
  bool removed = 4;
  google.protobuf.Timestamp created = 5;
  google.protobuf.Timestamp updated = 6;
}

message OfficePayload {
  uint64 id = 1 [(validate.rules).uint64.gt = 0];
  string name = 2  [(validate.rules).string = {min_len: 2, max_len: 100}];
  string description = 3;
}

message OfficeEvent {
  uint64 id = 1 [(validate.rules).uint64.gt = 0];
  uint64 office_id = 2 [(validate.rules).uint64.gt = 0];
  uint64 status = 3;
  string type = 4 [(validate.rules).string = {min_len: 2, max_len: 100}];
  google.protobuf.Timestamp created = 5;
  google.protobuf.Timestamp updated = 6;
  OfficePayload payload = 7;
}

message DescribeOfficeV1Request {
  uint64 office_id = 1 [(validate.rules).uint64.gt = 0];
}

message DescribeOfficeV1Response {
  Office value = 1;
}

message CreateOfficeV1Request {
  string name = 1  [(validate.rules).string = {min_len: 2, max_len: 100}];
  string description = 2;
}

message CreateOfficeV1Response {
  uint64 office_id = 1 [(validate.rules).uint64.gt = 0];
}

message ListOfficesV1Request {
  uint64 limit = 1 [(validate.rules).uint64 = {gt: 0, lt: 100}];
  uint64 offset = 2;
}

message ListOfficesV1Response {
  repeated Office items = 1;
}

message RemoveOfficeV1Request {
  uint64 office_id = 1 [(validate.rules).uint64.gt = 0];
}

message RemoveOfficeV1Response {
  bool found = 1;
}

message UpdateOfficeV1Request {
  uint64 office_id = 1 [(validate.rules).uint64.gt = 0];
  string name = 2 [(validate.rules).string={ min_len: 2, max_len: 100 }];
  string description = 3;
}

message UpdateOfficeV1Response {
  bool status = 1;
}